LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

IntegralProgrammer opened the pull request:
This Pull Request implements LFS-1059 by creating the `configure_ldap.py` program which configures a running CARDS instance to use an LDAP server for user authentication.

To test:

1. Build the `LFS-1059` branch
2. Download the test LDAP server (`docker pull dwimberger/ldap-ad-it`)
3. Start the test LDAP server (`docker run --rm -p 10389:10389 -it dwimberger/ldap-ad-it`)
4. Start CARDS
5. Attempt to login to CARDS with _Username:_ `test` and _Password:_ `secret`. This should fail and display an _unable to login_ message after approximately 1 minute.
6. Enter the `Utilities/Administration` directory
7. Run `python3 configure_ldap.py` and specify the following values at the following prompts:
    - `LDAP server hostname` = `localhost`
    - `LDAP server port` = `10389`
    - `LDAP Bind DN` = `uid=admin,ou=system`
    - `LDAP Bind Password` = `secret`
    - `LDAP User Base DN` = `dc=wimpi,dc=net`
    - `LDAP User Object Class` = `inetOrgPerson`
    - `LDAP User ID Attribute` = `samaccountname`

Alternatively, these values should be able to be specified through the `CARDS_LDAP_HOSTNAME`, `CARDS_LDAP_PORT`, `CARDS_LDAP_BIND_DN`, `CARDS_LDAP_BIND_PASSWORD`, `CARDS_LDAP_USER_BASE_DN`, `CARDS_LDAP_USER_OBJECT_CLASS`, and `CARDS_LDAP_USER_ID_ATTRIBUTE` environment variables.

8. You should now be able to log into CARDS with the _Username:_ `test` and _Password:_ `secret`

IntegralProgrammer committed:
LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

acrowthe reviewed (approved):
Tested: Both environment variables, runtime prompts and a mix work.

sdumitriu commented on Utilities/Administration/configure_ldap.py:
```diff
@@ -0,0 +1,155 @@
+if "CARDS_LDAP_PORT" in os.environ:
+  CARDS_LDAP_PORT = os.environ["CARDS_LDAP_PORT"]
+else:
+  CARDS_LDAP_PORT = input("LDAP server port: ")
```
I think this is not mandatory, the default (and almost always the one in use) is good enough. Make it optional, set only if there's an envvar for it.

IntegralProgrammer replied:
Implemented with commit `b23e2f2` :heavy_check_mark:

sdumitriu commented on Utilities/Administration/configure_ldap.py:
```diff
@@ -0,0 +1,155 @@
+form_data['user.idAttribute'] = CARDS_LDAP_USER_ID_ATTRIBUTE
+form_data['user.extraFilter'] = ''
+form_data['user.makeDnPath'] = 'false'
+form_data['group.baseDN'] = 'ou=groups,o=example,dc=com'
```
Why? This looks like something that won't match production instances, better to not send it at all if it is not specified in the env.

IntegralProgrammer replied:
Implemented with commit `aea91a8` :heavy_check_mark:

sdumitriu commented on Utilities/Administration/configure_ldap.py:
```diff
@@ -0,0 +1,155 @@
+form_data['noCertCheck'] = CARDS_LDAP_NO_CERT_CHECK
+form_data['bind.dn'] = CARDS_LDAP_BIND_DN
+form_data['bind.password'] = CARDS_LDAP_BIND_PASSWORD
+form_data['searchTimeout'] = '60s'
+form_data['adminPool.maxActive'] = '8'
+form_data['adminPool.lookupOnValidate'] = 'true'
+form_data['adminPool.lookupOnValidate'] = 'false'
+form_data['userPool.maxActive'] = '8'
+form_data['userPool.lookupOnValidate'] = 'true'
+form_data['userPool.lookupOnValidate'] = 'false'
```
These are the default values, which don't need to be specified. I think it would be better to only send what is explicitly configured, either by input (the mandatory properties), or by env vars.

These are very technical settings that probably don't need to be customized.

IntegralProgrammer replied:
Implemented with commit `aea91a8` :heavy_check_mark:

sdumitriu commented on Utilities/Administration/configure_ldap.py:
```diff
@@ -0,0 +1,155 @@
+form_data['user.baseDN'] = CARDS_LDAP_USER_BASE_DN
+form_data['user.objectclass'] = CARDS_LDAP_USER_OBJECT_CLASS
+form_data['user.idAttribute'] = CARDS_LDAP_USER_ID_ATTRIBUTE
+form_data['user.extraFilter'] = ''
+form_data['user.makeDnPath'] = 'false'
+form_data['group.baseDN'] = 'ou=groups,o=example,dc=com'
+form_data['group.objectclass'] = 'groupOfUniqueNames'
+form_data['group.nameAttribute'] = 'cn'
+form_data['group.extraFilter'] = ''
+form_data['group.makeDnPath'] = 'false'
+form_data['group.memberAttribute'] = 'uniquemember'
+form_data['useUidForExtId'] = 'false'
+form_data['customattributes'] = ''
```
I think it's better to not send these as well, if needed, they can be configured in the running instance.

IntegralProgrammer replied:
Implemented with commit `aea91a8` :heavy_check_mark:

sdumitriu reviewed (changes requested):
I would like it to have support for `--help`.

IntegralProgrammer committed:
LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

Use TCP port 389 for connection to LDAP server unless otherwise specified by environment variable

IntegralProgrammer committed:
LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

Do not send default configuration values for LDAP settings. Only send the variables that need to be set

IntegralProgrammer committed:
LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

configure_ldap.py will display a help screen if called with the --help flag

IntegralProgrammer commented:
> I would like it to have support for `--help`.

Implemented with commit `e6cd6bc` :heavy_check_mark:

sdumitriu commented on Utilities/Administration/configure_ldap.py:
```diff
@@ -0,0 +1,144 @@
+  print("- CARDS_LDAP_BIND_PASSWORD: Password for connecting to the LDAP server. Prompts if not specified.")
+  print("- CARDS_LDAP_USER_BASE_DN: Where to begin searching through LDAP for a user. Prompts if not specified.")
+  print("- CARDS_LDAP_USER_OBJECT_CLASS: The LDAP object type which will represent a CARDS user. Prompts if not specified.")
+  print("- CARDS_LDAP_USER_ID_ATTRIBUTE: The LDAP attribute that should match the login name.")
```
```suggestion
  print("- CARDS_LDAP_USER_ID_ATTRIBUTE: The LDAP attribute that should match the login name. Prompts if not specified.")
```

sdumitriu committed:
LFS-1059: Create a utility program for configuring LDAP settings for a running CARDS instance

Minor --help improvement.
